AnimeStandardFormatSearch option also looks for season packs

ttran913 opened the pull request:
#### Database Migration
NO

#### Description
Implemented the new search criteria object, `AnimeSeasonSearchCriteria`, and applied it where it might be needed throughout the codebase. Generally paired with wherever `AnimeEpisodeSearchCriteria` is referenced. Added some minor tests that are similar to the already existing tests. Now, when the `AnimeStandardFormatSearch` option is enabled in the indexer settings, season packs will also be found in addition to its original function.


#### Todos
- [x] Tests
- [ ] Wiki Updates


#### Issues Fixed or Closed by this PR

* https://github.com/Sonarr/Sonarr/issues/5616

markus101 commented on src/NzbDrone.Core/Indexers/Newznab/NewznabRequestGenerator.cs:
```diff
@@ -404,17 +404,45 @@ public virtual IndexerPageableRequestChain GetSearchRequests(AnimeEpisodeSearchC
                         $"&q={NewsnabifyTitle(queryTitle)}+{searchCriteria.AbsoluteEpisodeNumber:00}"));
 
-                    if (Settings.AnimeStandardFormatSearch && searchCriteria.SeasonNumber > 0 && searchCriteria.EpisodeNumber > 0)
+                    if (Settings.AnimeStandardFormatSearch && searchCriteria.SeasonNumber > 0)
```
A separate search is required, right now this will get called for every episode in the season.

ttran913 replied:
I made it so that it does another search separately in `ReleaseSearchService` for the season alone.

markus101 commented on src/NzbDrone.Core/Indexers/Newznab/NewznabRequestGenerator.cs:
```diff
@@ -404,17 +404,45 @@ public virtual IndexerPageableRequestChain GetSearchRequests(AnimeEpisodeSearchC
-                            $"&q={NewsnabifyTitle(queryTitle)}&season={NewznabifySeasonNumber(searchCriteria.SeasonNumber)}&ep={searchCriteria.EpisodeNumber}"));
+                        if (searchCriteria.IsSeasonSearch)
+                        {
+                            if (searchCriteria.SeasonNumber < 10)
```
With the `season` query param Sonarr doesn't need to pad the season number, the indexer should handle it correctly regardless and in the case of Prowlarr and Jackett they know best how to query the different indexers. This is the same as `ep`, which isn't padded.

ttran913 replied:
We removed on what we think the unnecessary code was since the indexers should already know best.

markus101 commented:
In order to get the latest changes you need to rebase on `develop`, instead of merging develop into your branch. Depending how you have your git setup, it'll be something like `git rebase upstream/develop`.

ttran913 commented:
We think we made some successful modifications. We hope we followed your suggestions sufficiently?

ttran913 commented:
Okay, we made so that it searches for the season pack only on the search of the first episode when conducting a whole season search. It will, of course, _not_ look for season packs when doing the individual search on the first episode.

markus101 reviewed (changes requested):
Instead of only handling this when searching for the first episode a separate search should be triggered to the indexer.

This will require a new search criteria, `AnimeSeasonSearchCriteria` and handling in each indexer.

ttran913 commented:
We implemented the new search criteria object, `AnimeSeasonSearchCriteria`, and implemented it where it might be needed throughout the codebase. We also added some minor tests that are similar to the already existing tests.

markus101 commented on src/NzbDrone.Core/Indexers/Newznab/NewznabRequestGenerator.cs:
```diff
@@ -422,6 +422,33 @@ public virtual IndexerPageableRequestChain GetSearchRequests(AnimeEpisodeSearchC
+                {
+                    if (Settings.AnimeStandardFormatSearch && searchCriteria.SeasonNumber > 0)
+                    {
+                        AddTvIdPageableRequests(pageableRequests,
```
This needs to be moved outside of the loop, right now it's triggered for each query title instead of just once. Looks like the original search for anime episodes does this as well, I'll fix those up.

markus101 commented:
Tested well except for the duplicate ID searches, just some things to address.

ttran913 commented on src/NzbDrone.Core/Indexers/BroadcastheNet/BroadcastheNetRequestGenerator.cs:
```diff
@@ -173,6 +173,32 @@ public virtual IndexerPageableRequestChain GetSearchRequests(AnimeEpisodeSearchC
+        {
+            var pageableRequests = new IndexerPageableRequestChain();
+
+            var parameters = new BroadcastheNetTorrentQuery();
```
All our formatting here and below was due to the fact that we were just trying to match style of the rest of the file.

ttran913 commented:
> Tested well except for the duplicate ID searches, just some things to address.

@markus101 How did you test for that? Was it in the testing files?

markus101 reviewed (approved):
Thanks for the fixes and cleanup, working as expected now.
